### PR TITLE
[Backport] Track-gen matching passing through TrackingParticle (11_2_X)

### PR DESCRIPTION
**Pull request by elusian**
#### PR description:
This PR adds a way to match `generalTracks` (in AOD) and `packedPFCandidates` and `lostTracks` (in miniAOD) to a subset of  `genParticles`, using association by hit on `TrackingParticles`.
The association is performed on step3 and just replicated in the miniAOD.
Since the track association by hit needs both `TrackingParticles` and `SimLink` for pixel and strips and the `SimLink` collections are not available in RAWSIM, two plugins are created to be run in step2 to add two pruned collections and avoid adding the all the `SimLink`, which have a non negligible size.

The changes contained are:
* a plugin for pruning `TrackingParticles` in step2 based on their association to specific `GenParticles`. The genParticle selection is taken from the `GenParticlePruner` class and can be configured.
* a plugin for pruning `SimLink` collections based on the `SimTrack` association to a `TrackingParticle`
* a plugin for adding an association between `PackedCandidates` and `GenParticles` based on an existing association between tracks and `GenParticles`
* Needed configuration changes to add this to standard sequences, gated behind the bParking modifier
* bugfix from #34113 

No plugin is needed for association between tracks and `GenParticles` through `TrackingParticle` as it is already present in CMSSW (`MCTrackMatcher`).

#### PR validation:
On a B+->Psi2SK sample (one of the possible targets for this PR) we registered this increases in size, timing and memory

  | Size | Timing | Memory
-- | -- | -- | --
RAWSIM | +0.05% | +0.08% | +5MB
RECOSIM | < +0.01% | +0.07% | +2MB
AODSIM | +0.06% |   |  
MINIAODSIM | +0.4% | +0.08% | +400kB

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
Backport of #33774  and #34113 for an Hackaton for GNN tracking

### CODE REVIEW

**Comment by mmusich** [comment]
please test

**Comment by slava77** [comment]
backport of #33774

File: SimTracker/TrackAssociation/plugins/DigiSimLinkPruner.cc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+#include "SimDataFormats/TrackerDigiSimLink/interface/PixelDigiSimLink.h"
+#include "SimDataFormats/TrackerDigiSimLink/interface/StripDigiSimLink.h"
+
+#include <unordered_set>
+
+namespace {
+
+  template <typename T>
+  std::vector<edm::DetSet<T>> pruneByTpAssoc(const edm::DetSetVector<T>& simLinkColl,
+                                             const std::set<std::pair<uint32_t, EncodedEventId>>& selectedIds) {
```
Comment (jpata) [inline SimTracker/TrackAssociation/plugins/DigiSimLinkPruner.cc:43]: Can you remind why std::set is used here, while std::unordered_set in master?
https://github.com/cms-sw/cmssw/blob/master/SimTracker/TrackAssociation/plugins/DigiSimLinkPruner.cc#L45

**Review by jpata — COMMENTED** [review]
Could you comment if this PR was tested in a bParking workflow in 11_2_X (1304.182)? Thanks!

File: SimTracker/TrackAssociation/plugins/DigiSimLinkPruner.cc (hunk 1 above)
Comment (elusian) [reply SimTracker/TrackAssociation/plugins/DigiSimLinkPruner.cc:43]: The ability to use std::unordered_set comes from #33462 which added the UniqueSimTrackId type and its hasher that enables std::unordered_set. It is not available on 11_2_X as far as I know. I had to do the same modification when backporting to 10_6_X.
TrackerHitAssociation/plugins/ClusterTPAssociationProducer.cc for example shows the same difference between 11_2_X and master

**Comment by elusian** [comment]
Yes, this was tested on 1304.182

**Comment by jpata** [comment]
+reconstruction
- backport of  #33774 and #34113
- no changes in reco outputs, the modifications are protected by the bParking modifier
- per https://github.com/cms-sw/cmssw/pull/34574#issuecomment-887328936, the bParking workflow was tested in 11_2_X

**Comment by civanch** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

**Comment by slava77** [comment]
@cms-sw/pdmv-l2 
please check this backport PR.
Thank you.

**Comment by kskovpen** [comment]
+1

**Comment by qliphy** [comment]
+1

**Comment by jpata** [comment]
This was merged a while ago, but is not in a release, as 11_2_X is closed. @elusian can you confirm that this is fine, or is something else required here?

**Comment by elusian** [comment]
This backport was for a Hackaton about GNN for tracking that concluded today. I believe the code was manually imported by the participants even if it was not in the release. No need to do anything else.

**Pull request merged**